match_merger.py

```diff
@@ -28,9 +29,17 @@ def merge(self, docs_matrix: List[DocumentArray], parameters: dict, **kwargs):
             self._merge_shard(results, docs, traversal_paths)
         return DocumentArray(list(results.values()))
 
+    def get_mean_score(self, docs):
+        # Any other non-ugly way to get the active metric?
+        metric_name = list(docs[0].scores.keys())[0]
+        return np.mean([doc.scores[metric_name].value for doc in docs])
+
     def _merge_shard(self, results, docs, traversal_paths):
         for doc in docs.traverse_flat(traversal_paths):
             if doc.id in results:
-                results[doc.id].matches.extend(doc.matches)
+                cur_mean = self.get_mean_score(results[doc.id].matches)
+                new_mean = self.get_mean_score(doc.matches)
+                if new_mean > cur_mean:
+                    results[doc.id] = doc
```

Comment:
I think this is not the right way to do it. Your logic choses the best match set. Imagine the following example (in this case: higher score = better):
```
a = Document(matches=[Document(scores={'my_score': 1}), Document(scores={'my_score': 6})])
b = Document(matches=[Document(scores={'my_score': 2}), Document(scores={'my_score': 4})])

merged(a,b, top_k=2)
expected result => Document(matches=[Document(scores={'my_score': 6}), Document(scores={'my_score': 4})])

your result => Document(matches=[Document(scores={'my_score': 1}), Document(scores={'my_score': 6})])
```

Reply:
You should choose per document the best `top_k` matches in all the `shards * top_k` matches.

Reply:
I am not sure if I understand this correctly but I think there might be a risk of duplicates. Here's what I think. 

Given that `a` and `b` in the above example are different shard results.

**Query:** What is 42?
**Shard-a:** 
`[('42 is the answer to everything', score=1.0), ('42 is a real number', score=0.8), ('42 is an even number', score=0.6)]`
**Shard-b:** 
`[('42 is the answer to everything', score=0.95), ('42 is a real number', score=0.8), ('42 is an even number', score=0.7)]`
**Shard-c:** 
`[('42 is the answer to everything', score=0.99), ('42 is a real number', score=0.6), ('42 is an even number', score=0.7)]`

**After MatchMerger**
```
merged(a, b, c, top_k=5)
[('42 is the answer to everything', score=1.0), 
('42 is the answer to everything', score=0.99), 
('42 is the answer to everything', score=0.95), 
('42 is a real number', score=0.8),
('42 is a real number', score=0.8)]
```

Let me know what you think.

Thanks :)

Reply:
Document duplication concern should not be put in this executor. Sharding sends documents to a single shard so it would not exist if the user does not provide this twice. 

And if the user provides this twice is the responsability of the Flow designer and the client to take care of